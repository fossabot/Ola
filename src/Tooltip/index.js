import React, { useState, useRef } from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'
import useOutsideEvent from '../hooks/useOutsideEvent'

const getClassPosition = ref => {

  const win_width = window.innerWidth
  const win_height = window.innerHeight

  const { top, left, width, height } = ref.current.getClientRects()[0]
  const el_left = left + (width / 2)
  const el_top = top + (height / 2)

  const heightPosition = ((win_height * 0.5) < el_top ) ? 'top' : 'bottom'
  const widthPosition = ((win_width * 0.33) > el_left) ? 'right' : ((win_width * 0.66) > el_left) ? 'center' : 'left'

  return `${heightPosition}${widthPosition}`
}

const Tooltip = ({ trigger, children }) => {

  const [position, setPosition] = useState(null)
  const tooltipRef = useRef(null)

  useOutsideEvent(tooltipRef, ref => {
    ref.current.open = false
    setPosition(null)
  })

  const toggle = () => tooltipRef.current.open ? setPosition( getClassPosition(tooltipRef) ) : setPosition(null)

  return (
    <details className={cx('ola_tooltip')} onToggle={toggle} ref={tooltipRef}>
      <summary className="ola_tooltip-trigger">{trigger}</summary>
      <div className={cx('ola_tooltip-content', position && `is-${position}`)}>{children}</div>
    </details>
  )
}

Tooltip.propTypes = {
  /** Trigger nodes */
  trigger: PT.oneOfType([
    PT.string,
    PT.arrayOf(PT.node),
    PT.node
  ]).isRequired,
  /** Childen nodes */
  children: PT.oneOfType([
    PT.string,
    PT.arrayOf(PT.node),
    PT.node
  ]).isRequired
}

export default Tooltip
