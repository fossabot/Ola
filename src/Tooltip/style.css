.ola_tooltip {
    display: inline-block;
    position: relative;
}
.ola_tooltip-trigger {
    list-style: none;
    outline: 0;
    &::-webkit-details-marker {
        display: none;
    }
}
.ola_tooltip-content {
    will-change: opacity, transform;
    --size: var(--size-5);
    position: absolute;
    filter: drop-shadow(var(--shadow-2)) drop-shadow(var(--shadow-line));
    background: var(--white);
    border-radius: var(--radius);
    padding: var(--size-6);
    width: var(--size-11);
    box-sizing: border-box;
    z-index: 5;

    &::before {
        content: "";
        display: block;
        position: absolute;
    }

    &.is-bottomright,
    &.is-bottomcenter,
    &.is-bottomleft {
        margin-top: var(--size-3);

        &::before {
            border-bottom: solid var(--size) white;
            border-right: solid var(--size) transparent;
            border-left: solid var(--size) transparent;
            top: calc(var(--size) * -1);
        }
    }
    &.is-topright,
    &.is-topcenter,
    &.is-topleft {
        bottom: 100%;
        margin-bottom: var(--size-3);

        &::before {
            border-top: solid var(--size) white;
            border-right: solid var(--size) transparent;
            border-left: solid var(--size) transparent;
            bottom: calc(var(--size) * -1);
        }
    }

    &.is-topright,
    &.is-bottomright {
        left: calc(50% - var(--size) * 2);

        &::before {
            left: var(--size);
        }
    }

    &.is-topleft,
    &.is-bottomleft {
        right: calc(50% - var(--size) * 2);

        &::before {
            right: var(--size);
        }
    }
    &.is-topcenter,
    &.is-bottomcenter {
        left: calc(50% - var(--size-11) / 2);

        &::before {
            right: calc(50% - var(--size));
        }
    }
}
.ola_tooltip[open] {
    & .ola_tooltip-trigger .ola_buttonIcon,
    & .ola_tooltip-trigger .ola_button {
        color: var(--color-hover);
        background-color: var(--background-hover);
    }
    &.ola_tooltip-content {
        animation: tooltip-show 200ms;
    }
}

@keyframes tooltip-show {
    from {
        opacity: 0;
        transform: translateY(10px);
    }
    50% {
        opacity: 1;
    }
    to {
        transform: translateY(0);
    }
}
