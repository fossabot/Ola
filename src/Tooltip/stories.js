import React from 'react'
import { storiesOf } from '@storybook/react'
import { radios } from '@storybook/addon-knobs'

import Tooltip from './'
import Button from '../Button'
import Icon from '../Icon'
import ButtonIcon from '../ButtonIcon'
import Table from '../Table'
import TableCell from '../Table/Cell'
import Modal from '../Modal'
import ModalHeader from '../Modal/Header'
import ModalContent from '../Modal/Content'

storiesOf('Tooltip')
  .add('Default', () => (
    <figure>
      <Tooltip trigger={<p>Click for show content</p>}>Content</Tooltip>
    </figure>
  ))
  .add('With Button', () => (
    <figure>
      <Tooltip trigger={<Button as="span" variant="primary">Trigger as span</Button>}>Lorem ipsum</Tooltip>
    </figure>
  ))
  .add('With ButtonIcon', () => (
    <figure>
      <Tooltip trigger={<ButtonIcon as="span"><Icon name="help" /></ButtonIcon>}>Lorem ipsum</Tooltip>
    </figure>
  ))
  .add('Within a scrollable element', () => (
    <figure>
      <div style={{height: '300px', width: '300px', overflow: 'auto', padding: '1em', border: 'solid'}}>
        <div style={{padding: '250px'}}>
          <Tooltip trigger={<ButtonIcon as="span"><Icon name="help" /></ButtonIcon>}>Lorem ipsum</Tooltip>
        </div>
      </div>
    </figure>
  ))
  .add('Variants', () => (
    <figure>
      <Tooltip variant={radios('Variant', ['', 'wide'], '')} trigger={<ButtonIcon as="span"><Icon name="help" /></ButtonIcon>}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        Quisque ligula lacus, faucibus vitae sodales a, eleifend in massa.
        Donec sapien massa, interdum et facilisis vitae, elementum ac orci.
        Nam id blandit turpis. Cras eleifend accumsan nisi vitae luctus.
        Donec ultricies auctor sodales.
      </Tooltip>
    </figure>
  ))
  .add('In a table', () => (
    <figure>
      <Modal open>
        <ModalHeader
          title="Modal Header"
          intro={<><strong>Lorem ipsum</strong> for testing intro</>} />
        <ModalContent variant="scroll">
          <Table caption="Superheros and sidekicks" responsive>
            <thead>
              <tr>
                <TableCell header>
                  <span>Your current page title</span>
                  <Tooltip trigger={<ButtonIcon as="span" variant="secondary"><Icon name="help" size="small" /></ButtonIcon>}>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</Tooltip>
                </TableCell>
                <TableCell header variant="right">
                  <span>Incoming links</span>
                  <Tooltip trigger={<ButtonIcon as="span" variant="secondary"><Icon name="help" size="small" /></ButtonIcon>}>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</Tooltip>
                </TableCell>
                <TableCell header variant="center">
                  <span>Popularity</span>
                  <Tooltip trigger={<ButtonIcon as="span" variant="secondary"><Icon name="help" size="small" /></ButtonIcon>}>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</Tooltip>
                </TableCell>
              </tr>
            </thead>
            <tbody>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
              <tr>
                <TableCell>value</TableCell>
                <TableCell variant="right">value</TableCell>
                <TableCell variant="center">value</TableCell>
              </tr>
            </tbody>
          </Table>
        </ModalContent>
      </Modal>
    </figure>
  ))
