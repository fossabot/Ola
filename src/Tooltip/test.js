import React from 'react'
import Tooltip from './'
import renderer from 'react-test-renderer'

it('Default Tooltip', () => {
  const tree = renderer.create(<Tooltip trigger="Trigger">Content</Tooltip>).toJSON()
  expect(tree).toMatchSnapshot()
})

it('Tooltip with nodes', () => {
  const tree = renderer.create(<Tooltip trigger={<p>Node trigger</p>}><p>Node content</p></Tooltip>).toJSON()
  expect(tree).toMatchSnapshot()
})
