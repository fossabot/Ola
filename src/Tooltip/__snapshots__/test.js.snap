// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Tooltip 1`] = `
<details
  className="ola_tooltip"
  onToggle={[Function]}
>
  <summary
    className="ola_tooltip-trigger"
  >
    Trigger
  </summary>
  <div
    className="ola_tooltip-content"
  >
    Content
  </div>
</details>
`;

exports[`Tooltip with nodes 1`] = `
<details
  className="ola_tooltip"
  onToggle={[Function]}
>
  <summary
    className="ola_tooltip-trigger"
  >
    <p>
      Node trigger
    </p>
  </summary>
  <div
    className="ola_tooltip-content"
  >
    <p>
      Node content
    </p>
  </div>
</details>
`;
